clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
+        ) => {
+            let contract_to_check = db
+                .load_contract(&contract_identifier)
+                .ok_or(CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+            runtime_cost(ClarityCostFunction::AnalysisFetchContractEntry, tracker, 1)?;
+            let expected_trait = lookup_trait(
+                db,
+                contract_context,
+                expected_trait_id,
+                clarity_version,
+                tracker,
+            )?;
+            contract_to_check.check_trait_compliance(expected_trait_id, &expected_trait)?;
+        }
+        (
+            TypeSignature::ListUnionType(types),
+            TypeSignature::CallableType(CallableSubtype::Trait(_)),
+        ) => {
+            // Verify that all types in the union implement this trait
+            for subtype in types {
+                inner_type_check_type(
+                    db,
+                    contract_context,
+                    &TypeSignature::CallableType(subtype.clone()),
+                    expected_type,
+                    clarity_version,
+                    depth + 1,
+                    tracker,
+                )?;
+            }
+        }
+        (TypeSignature::NoType, _) => (),
+        (_, _) => {
+            if !expected_type.admits_type(&actual_type) {
+                return Err(
+                    CheckErrors::TypeError(expected_type.clone(), actual_type.clone()).into(),
+                );
+            }
+        }
+    }
+    Ok(expected_type.clone())
+}
+
+fn lookup_trait<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    trait_id: &TraitIdentifier,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> CheckResult<BTreeMap<ClarityName, FunctionSignature>> {
+    if let Some(contract_context) = contract_context {
+        // If the trait is from this contract, then it must be in the context or it doesn't exist.
+        if contract_context.is_contract(&trait_id.contract_identifier) {
+            return Ok(if clarity_version < ClarityVersion::Clarity2 {
+                contract_context.get_trait(&trait_id.name)
+            } else {
+                contract_context.get_trait_by_id(trait_id)
+            }
+            .ok_or(CheckErrors::NoSuchTrait(
+                trait_id.contract_identifier.to_string(),
+                trait_id.name.to_string(),
+            ))?
+            .clone());
+        }
+        if clarity_version >= ClarityVersion::Clarity2 {
+            if let Some(trait_sig) = contract_context.get_trait_by_id(trait_id) {
+                return Ok(trait_sig.clone());
+            }
+        }
+    }
+
+    match db
+        .get_defined_trait(&trait_id.contract_identifier, &trait_id.name)?
+        .ok_or(
+            CheckErrors::NoSuchTrait(
+                trait_id.contract_identifier.to_string(),
+                trait_id.name.to_string(),
+            )
+            .into(),
```

Comment:
Does this error path need a runtime cost?  I see below that we're applying trivial-sized runtime costs for `CheckErrors` cases, but the absence of a runtime cost here is inconsistent.

It's tricky to assess a runtime cost on a `CheckErrors` error path, because usually it means that a database query simply failed (and we don't have an easy way of telling what the "true" cost in the DB is for a negative result).  Also, arguably transactions that encounter `CheckErrors` shouldn't be mined in the first place, but we are not yet certain that this is 100% true (which is an argument *for* runtime costs on all error paths).

What do you think?

Reply:
I agree, it's better to include the cost on all error paths. The missing costs in this PR are a combination of me copying the old implementation, which was missing the costs in many places, and me struggling to understand exactly what the goals are with the costs. I think I have a better understanding now and I will add these missing costs.

Reply:
Added in 90c68ffbf